backend/de.metas.async/src/main/java/de/metas/async/processor/impl/planner/QueueProcessorPlanner.java

```diff
@@ -0,0 +1,360 @@
+			throw new AdempiereException("Planner is not running! Hint: QueueProcessorPlanner should be started via start() method!");
+		}
+
+		while (isRunning.get())
+		{
+			try
+			{
+				final boolean successfulRun = run0();
+
+				if (!successfulRun && stopOnFailedRun())
+				{
+					logger.warn("*** QueueProcessorPlanner.run() -> last run was not successful & planner = " + this.getClass().getName() + " has stopOnFailedRun policy; Shutting down...");
+					shutdown();
+				}
+				else
+				{
+					// note: we always get the new service, because things might have changed since this method started
+					final int pollIntervalMs = Services.get(ISysConfigBL.class).getIntValue(SYSCONFIG_POLLINTERVAL, 1000);
+
+					Thread.sleep(pollIntervalMs);
+				}
+			}
+			catch (final InterruptedException e)
+			{
+				logger.error("Got interrupt request.", e);
+				return;
+			}
+			catch (final Throwable throwable)
+			{
+				logger.warn("*** QueueProcessorPlanner.run() caught an unexpected error!", throwable);
+
+				if (stopOnFailedRun())
+				{
+					logger.error("*** QueueProcessorPlanner.run() caught an unexpected error!"
+										 + " Planner = " + this.getClass().getName() + " has stopOnFailedRun policy; Shutting down...", throwable);
+					shutdown();
+
+					throw AdempiereException.wrapIfNeeded(throwable);
+				}
+			}
+		}
+	}
+
+	@NonNull
+	public List<IQueueProcessor> getRegisteredQueueProcessors()
+	{
+		return ImmutableList.copyOf(queueProcessors.values());
+	}
+
+	public void addQueueProcessor(@NonNull final IQueueProcessor queueProcessor)
+	{
+		validateQueueProcessorAssignment(queueProcessor);
+
+		queueProcessors.put(queueProcessor.getQueueProcessorId(), queueProcessor);
+	}
+
+	public boolean removeQueueProcessor(@NonNull final QueueProcessorId queueProcessorId)
+	{
+		final Optional<IQueueProcessor> processorOpt = getQueueProcessor(queueProcessorId);
+
+		if (!processorOpt.isPresent())
+		{
+			return true;
+		}
+
+		processorOpt.get().shutdownExecutor();
+
+		queueProcessors.remove(processorOpt.get().getQueueProcessorId());
+
+		return true;
+	}
+
+	@NonNull
+	public Optional<IQueueProcessor> getQueueProcessor(@NonNull final QueueProcessorId queueProcessorId)
+	{
+		return Optional.ofNullable(queueProcessors.get(queueProcessorId));
+	}
+
+	public void start()
+	{
+		mainLock.lock();
+		try
+		{
+			if (isRunning.get())
+			{
+				logger.debug("QueueProcessorPlanner.start() - already running! return...");
+				return;
+			}
+
+			this.startPlanner();
+		}
+		finally
+		{
+			mainLock.unlock();
+		}
+	}
+
+	public void shutdown()
+	{
+		mainLock.lock();
+		try
+		{
+			if (!queueProcessors.isEmpty())
+			{
+				queueProcessors
+						.values()
+						.stream()
+						.map(IQueueProcessor::getQueueProcessorId)
+						.forEach(this::removeQueueProcessor);
+			}
+
+			this.shutdownPlanner();
+		}
+		finally
+		{
+			mainLock.unlock();
+		}
+	}
+
+	private boolean run0()
+	{
+		final List<IQueueProcessor> availableProcessors = queueProcessors
```

Comment:
consider extracting this to a method called getQueueProcessorsAvailableToWork().